src/IIIFPresentation/Services/Manifests/ManifestItemsParser.cs

```diff
@@ -174,22 +177,38 @@ private CanvasPainting CreatePartialCanvasPainting(ResourceBase? resource,
         return cp;
     }
     
-    private string? TryGetValidCanvasId(int customerId, Canvas currentCanvas)
+    private string? TryGetValidCanvasId(int customerId, Canvas currentCanvas, 
+        List<CanvasPainting>? paintedResourceCanvasPaintings)
     {
         if (currentCanvas.Id == null) return null;
 
         if (!Uri.TryCreate(currentCanvas.Id, UriKind.Absolute, out var canvasId))
         {
-            CanvasHelper.CheckForProhibitedCharacters(currentCanvas.Id);
+            currentCanvas.Id = CanvasHelper.CheckForProhibitedCharacters(currentCanvas.Id, logger, false);
+            
+            if (currentCanvas.Id != null && !(paintedResourceCanvasPaintings?.Any(cp => cp.Id == currentCanvas.Id) ?? false))
+            {
+                throw new InvalidCanvasIdException(currentCanvas.Id, "The canvas id is not a valid URI, and cannot be matched with a painted resource");
+            }
+            
             return currentCanvas.Id;
         }
 
         if (IsRecognisedHost(customerId, canvasId.Host))
         {
             var parsedCanvasId =
                 pathRewriteParser.ParsePathWithRewrites(canvasId.Host, canvasId.AbsolutePath, customerId);
-            CanvasHelper.CheckParsedCanvasIdForErrors(parsedCanvasId, canvasId.AbsolutePath);
-            return parsedCanvasId.Resource;
+
+            var checkedCanvasId =
+                CanvasHelper.CheckParsedCanvasIdForErrors(parsedCanvasId, canvasId.AbsolutePath, logger, false);
+            
+            // cannot match up with a painted resource, so don't set the canvas id
+            if (!(paintedResourceCanvasPaintings?.Any(cp => cp.Id == checkedCanvasId) ?? false))
```

Comment:
Small - would it be clearer if this returned val if we _can_ match, else let it fall through? (only mentioning as it took me a sec to understand when scanning code).

`if isMatching return match` generally reads better than `if not isMatching return nothing else return match`